### PR TITLE
Fix snapshot in streamKafka & test

### PR DESCRIPTION
**Pull request by viliam-durina**
Also includes some clean-up in ArrayDequeInbox

### CODE REVIEW

File: hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/impl/connector/kafka/StreamKafkaP.java [hunk 1]
```diff
@@ -91,36 +96,45 @@
...
             List<PartitionInfo> partitionInfos = consumer.partitionsFor(topicName);
-            getLogger().finest("Num of ptions for topic " + topicName + ": " + partitionInfos.size());
-            while (nextAssignablePtions[topicIdx] < partitionInfos.size()) {
-                int partition = nextAssignablePtions[topicIdx];
-                nextAssignablePtions[topicIdx] += processorCount;
-                assert partitionInfos.get(partition).partition() == partition;
-                assignment.add(new TopicPartition(topicName, partition));
-                changed = true;
+            LoggingUtil.logFinest(getLogger(), "Num of ptions for topic '%s': %d", topicName, partitionInfos.size());
+            if (nextAssignablePtions[topicIdx] < partitionInfos.size()) {
```
Comment (mtopolnik) [inline hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/impl/connector/kafka/StreamKafkaP.java]: Here you can reduce the nesting level by inverting the `if` and using `continue`.

File: hazelcast-jet-kafka/src/test/java/com/hazelcast/jet/impl/connector/kafka/StreamKafkaPTest.java [hunk 2]
```diff
@@ -94,6 +150,111 @@ public void run() throws Exception {
         assertTrueEventually(() -> assertTrue(jobFuture.isDone()));
     }
 
+    @Test
+    public void when_snapshotSaved_then_offsetsRestored() throws Exception {
+        StreamKafkaP[] processor = {new StreamKafkaP(properties, singletonList(topic1Name), 1, 0, 60000)};
+        TestOutbox[] outbox = {new TestOutbox(new int[] {10}, 10)};
```
Comment (mtopolnik) [inline hazelcast-jet-kafka/src/test/java/com/hazelcast/jet/impl/connector/kafka/StreamKafkaPTest.java]: I can't make out a reason why these two are packed into an array.

File: hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/impl/connector/kafka/StreamKafkaP.java [hunk 3]
```diff
@@ -91,36 +96,45 @@
...
     }
 
     @Override
     public boolean complete() {
         if (System.nanoTime() >= nextPartitionCheck) {
-            reassignPartitions();
+            reassignPartitions(false);
         }
 
         if (!assignment.isEmpty()) {
```
Comment (mtopolnik) [inline hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/impl/connector/kafka/StreamKafkaP.java]: Here also the nesting level can be reduced by inverting the `if` and returning false after `parkNanos()`

File: hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/impl/connector/kafka/StreamKafkaP.java [hunk 4]
```diff
@@ -254,8 +272,8 @@ public MetaSupplier(Properties properties, List<String> topicIds) {
             } else {
                 metadataRefreshInterval = KAFKA_DEFAULT_REFRESH_INTERVAL;
             }
-            // Set metadata caching to 1 second: we get metadata for multiple partitions one by one, but
-            // internally it is fetched at once. This enables for the other calls to just query fetched metadata.
+            // Set metadata caching to 1 second: we read the metadata for multiple partitions one by one. If we
```
Comment (mtopolnik) [inline hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/impl/connector/kafka/StreamKafkaP.java]: We should consistently apply the 72-char text line length rule even to these comments.

File: hazelcast-jet-kafka/src/test/java/com/hazelcast/jet/impl/connector/kafka/StreamKafkaPTest.java (hunk 2 above)
Comment (viliam-durina) [reply hazelcast-jet-kafka/src/test/java/com/hazelcast/jet/impl/connector/kafka/StreamKafkaPTest.java]: They are both modified and used in lambdas.

**Comment by viliam-durina** [comment]
verify

**Comment by cangencer** [comment]
verify

File: hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/impl/connector/kafka/StreamKafkaP.java [hunk 5]
```diff
@@ -91,53 +96,64 @@
...
         }
 
-        if (changed) {
-            getLogger().info("Partition assignment changed: " + assignment);
+        if (!addedPartitions.isEmpty()) {
+            getLogger().info("Partition assignment changed, new partitions: " + addedPartitions);
+            assignment.addAll(addedPartitions);
             consumer.assign(assignment);
+            if (!isInitial) {
+                consumer.seekToBeginning(addedPartitions);
```
Comment (cangencer) [inline hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/impl/connector/kafka/StreamKafkaP.java]: does this call take `auto.offset.reset` into account?

File: hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/impl/connector/kafka/StreamKafkaP.java [hunk 6]
```diff
@@ -91,53 +96,64 @@
     protected void init(@Nonnull Context context) throws Exception {
         snapshottingEnabled = context.snapshottingEnabled();
         consumer = new KafkaConsumer<>(properties);
-        reassignPartitions();
+        reassignPartitions(true);
```
Comment (cangencer) [inline hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/impl/connector/kafka/StreamKafkaP.java]: wouldn't it make more sense to do partition assignment in the Supplier/MetaSupplier. It looks like we're relying on each node independently deciding what partitions it should subscribe to?

File: hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/impl/connector/kafka/StreamKafkaP.java [hunk 7]
```diff
@@ -155,17 +171,21 @@ public boolean isCooperative() {
     public boolean saveToSnapshot() {
         if (snapshotTraverser == null) {
             snapshotTraverser = Traversers.traverseIterable(offsets.entrySet())
-                    .onFirstNull(() -> snapshotTraverser = null);
+                                          .map(e -> entry(broadcastKey(e.getKey()), e.getValue()))
+                                          .onFirstNull(() -> snapshotTraverser = null);
         }
         return emitFromTraverserToSnapshot(snapshotTraverser);
```
Comment (cangencer) [inline hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/impl/connector/kafka/StreamKafkaP.java]: shouldn't the offsets be commited after they have been saved to snapshot?

File: hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/impl/connector/kafka/StreamKafkaP.java (hunk 6 above)
Comment (viliam-durina) [reply hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/impl/connector/kafka/StreamKafkaP.java]: 1. you only know the partitions when you have a `consumer`. You don't have one in Supplier. 
2. we regularly check for new partitions, kafka supports adding partitions during runtime (unlike HZ)

File: hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/impl/connector/kafka/StreamKafkaP.java (hunk 5 above)
Comment (viliam-durina) [reply hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/impl/connector/kafka/StreamKafkaP.java]: Intentionally not. This is for partitions added after the initial assignment: we want to read them from the beginning.

File: hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/impl/connector/kafka/StreamKafkaP.java (hunk 7 above)
Comment (viliam-durina) [reply hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/impl/connector/kafka/StreamKafkaP.java]: No need to: we don't use them. I commit them when snapshots are disabled to have "approximately once" guarantee.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/CloseableProcessorSupplier.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    @Override
+    public void complete(Throwable error) {
+        if (processors == null) {
+            return;
+        }
+
+        Throwable firstError = null;
+        // close all processors, ignoring their failures and throwing the first failure (if any)
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/CloseableProcessorSupplier.java]: This is a self-contradictory comment. If we throw, we don't ignore. I prefer to just delete the comment, the code speaks for itself.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlanBuilder.java [hunk 9]
```diff
@@ -78,7 +79,9 @@ private ExecutionPlanBuilder() {
             final List<EdgeDef> outbound = toEdgeDefs(dag.getOutboundEdges(vertex.getName()), defaultEdgeConfig,
                     e -> vertexIdMap.get(e.getDestName()), isJobDistributed);
             final ProcessorMetaSupplier metaSupplier = vertex.getSupplier();
-            metaSupplier.init(new MetaSupplierCtx(instance, totalParallelism, localParallelism,
+            ILogger logger = nodeEngine.getLogger(metaSupplier.getClass().getName() + "." + vertex.getName() +
+                    "#MetaProcessorSupplier");
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlanBuilder.java]: `#ProcessorMetaSupplier`

**Pull request merged**